src/chapters/10-rag.md

```diff
@@ -0,0 +1,658 @@
+
+![](../images/chapter-10/embedding-data-flow.png)
+
+### Indexer Flow Implementation
+
+The indexer flow coordinates the entire indexing process, orchestrating document processing, chunking, embedding generation, and storage. This flow represents the phase of RAG where you prepare your knowledge base.
+
+**The indexing process involves these steps:**
+
+1. **Document loading**: Read and validate the input document
+2. **Text extraction**: Convert document format to plain text
+3. **Chunking**: Split text into semantically meaningful pieces
+4. **Embedding generation**: Convert each chunk into a vector representation using an embedding model
+5. **Storage**: Save vectors and metadata to a searchable index
+
+Here's our complete indexer flow implementation:
+
+```go
+// IndexerRequest represents the input for the indexer flow
+type IndexerRequest struct {
+    PDFPath string `json:"pdfPath,omitempty"`
+}
+
+// NewIndexerFlow creates a flow that reads PDF documents, creates embeddings, and stores them in localvec.
+func NewIndexerFlow(g *genkit.Genkit, tools []ai.ToolRef, docStore *localvec.DocStore) *core.Flow[IndexerRequest, string, struct{}] {
+    return genkit.DefineFlow(g, "indexerFlow", func(ctx context.Context, req IndexerRequest) (string, error) {
+        // Default PDF path if not provided
+        pdfPath := req.PDFPath
+        if pdfPath == "" {
+            // Use absolute path to the Arduino report
+            pdfPath = "internal/docs/arduino_report.pdf"
+        }
+
+        // Make path absolute
+        absPath, err := filepath.Abs(pdfPath)
+        if err != nil {
+            return "", fmt.Errorf("failed to get absolute path: %w", err)
+        }
+
+        // Parse PDF into chunks
+        chunks, err := rag.ParsePDFToChunks(absPath, 1000) // 1000 character chunks
+        if err != nil {
+            return "", fmt.Errorf("failed to parse PDF: %w", err)
+        }
+
+        // Index the documents
+        err = localvec.Index(ctx, chunks, docStore)
+        if err != nil {
+            return "", fmt.Errorf("failed to index documents: %w", err)
+        }
+
+        return fmt.Sprintf("Successfully indexed %d chunks from %s", len(chunks), pdfPath), nil
+    })
+}
+```
+
+**What happens during `localvec.Index()`:**
+
+Behind the scenes, this function:
+
+- Takes each text chunk and converts it to a vector using the configured embedding model
+- Stores the vectors along with the original text and metadata in the local vector database
+- Creates search indexes for efficient similarity queries
+- Persists data to disk for future retrieval operations under `.genkit/indexes`
+
+
+### Testing the Indexer Flow
+
+You can test the indexer flow using the developer UI or curl:
+
+![](../images/chapter-10/indexer-ui.png)
+
+```bash
+# Index the default Arduino report
+curl -X POST http://localhost:9090/indexerFlow \
+  -H "Content-Type: application/json" \
+  -d '{}'
```

Comment:
Genkit HTTP handlers expect the request body to be wrapped in a `data` field.
`  -d '{"data":{}}'`